crates/viewer/re_arrow_ui/src/show_index.rs

```diff
@@ -0,0 +1,658 @@
+                let value = self.value(idx);
+                let s = $fmt(value);
+                f.code_primitive(&s);
+                Ok(())
+            }
+
+            fn array(&self) -> &dyn Array {
+                self
+            }
+        })+
+    };
+}
+
+primitive_display!(re_format::format_int: Int8Type, Int16Type, Int32Type, Int64Type);
+primitive_display!(re_format::format_uint: UInt8Type, UInt16Type, UInt32Type, UInt64Type);
+primitive_display!(re_format::format_f32: Float32Type);
+primitive_display!(re_format::format_f64: Float64Type);
+
+fn format_f16(value: f16) -> String {
+    re_format::format_f32(value.to_f32())
+}
+primitive_display!(format_f16: Float16Type);
+
+impl<OffsetSize: OffsetSizeTrait> ShowIndex for &GenericBinaryArray<OffsetSize> {
+    fn write(&self, idx: usize, f: &mut SyntaxHighlightedBuilder<'_>) -> EmptyArrowResult {
+        let value = self.value(idx);
+        f.code_primitive(&re_format::format_bytes(value.len() as f64));
+        Ok(())
+    }
+
+    fn array(&self) -> &dyn Array {
+        self
+    }
+}
+
+impl<'a, K: ArrowDictionaryKeyType> ShowIndexState<'a> for &'a DictionaryArray<K> {
+    type State = Box<dyn ShowIndex + 'a>;
+
+    fn prepare(&self, options: &FormatOptions<'a>) -> Result<Self::State, ArrowError> {
+        make_ui(self.values().as_ref(), options)
+    }
+
+    fn write(
+        &self,
+        s: &Self::State,
+        idx: usize,
+        f: &mut SyntaxHighlightedBuilder<'_>,
+    ) -> EmptyArrowResult {
+        let value_idx = self.keys().values()[idx].as_usize();
+        s.as_ref().write(value_idx, f)
+    }
+}
+
+impl<'a, K: RunEndIndexType> ShowIndexState<'a> for &'a RunArray<K> {
+    type State = Box<dyn ShowIndex + 'a>;
+
+    fn prepare(&self, options: &FormatOptions<'a>) -> Result<Self::State, ArrowError> {
+        make_ui(self.values().as_ref(), options)
+    }
+
+    fn write(
+        &self,
+        s: &Self::State,
+        idx: usize,
+        f: &mut SyntaxHighlightedBuilder<'_>,
+    ) -> EmptyArrowResult {
+        let value_idx = self.get_physical_index(idx);
+        s.as_ref().write(value_idx, f)
+    }
+}
+
+fn write_list(
+    f: &mut SyntaxHighlightedBuilder<'_>,
+    mut range: Range<usize>,
+    values: &dyn ShowIndex,
+) -> EmptyArrowResult {
+    const MAX_LIST_ITEMS: usize = 10;
+    f.code_syntax("[");
+    if let Some(idx) = range.next() {
+        values.write(idx, f)?;
+    }
+
+    let mut items = 1;
+
+    for idx in range {
+        if items >= MAX_LIST_ITEMS {
+            f.code_syntax(", ...");
+            break;
+        }
+        f.code_syntax(", ");
+        values.write(idx, f)?;
+        items += 1;
+    }
+    f.code_syntax("]");
+    Ok(())
+}
+
+/// Show a list.
+///
+/// If there are enough items, it will show items in a tree of ranges.
+/// Since arrow arrays might not start at 0, you can pass a `Range<usize>`.
+/// The indexes shown in the UI will be _normalized_ so it's always `0..end-start`
```

Comment:
Wait what, is not all arrow arrays zero-indexed? Can you link to some docs for this?